controllers/envoysecuritypolicy_referencegrant_controller.go

```diff
@@ -0,0 +1,160 @@
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *EnvoySecurityPolicyReferenceGrantReconciler) securityPolicyReferenceGrant(ctx context.Context, kuadrantNamespace string) (*gatewayapiv1beta1.ReferenceGrant, error) {
+	logger, _ := logr.FromContext(ctx)
+	logger = logger.WithName("securityPolicyReferenceGrant")
+
+	rg := &gatewayapiv1beta1.ReferenceGrant{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      KuadrantReferenceGrantName,
+			Namespace: kuadrantNamespace,
+		},
+		Spec: gatewayapiv1beta1.ReferenceGrantSpec{
+			To: []gatewayapiv1beta1.ReferenceGrantTo{
+				{
+					Group: "",
+					Kind:  "Service",
+					Name:  ptr.To[gatewayapiv1.ObjectName](kuadrant.AuthorinoServiceName),
+				},
+			},
+		},
+	}
+
+	espNamespaces := make(map[string]struct{})
+	listOptions := &client.ListOptions{LabelSelector: labels.SelectorFromSet(map[string]string{kuadrant.KuadrantNamespaceAnnotation: kuadrantNamespace})}
+	espList := &egv1alpha1.SecurityPolicyList{}
+	if err := r.Client().List(ctx, espList, listOptions); err != nil {
+		return nil, err
+	}
+
+	for _, esp := range espList.Items {
+		// only append namespaces that differ from the kuadrant namespace and are not marked for deletion
+		if esp.DeletionTimestamp == nil && esp.Namespace != kuadrantNamespace {
+			espNamespaces[esp.Namespace] = struct{}{}
+		}
+	}
+
+	if len(espNamespaces) == 0 {
+		logger.V(1).Info("no security policies exist outside of the kuadrant namespace, skipping ReferenceGrant")
+		utils.TagObjectToDelete(rg)
+		return rg, nil
+	}
+
+	refGrantFrom := lo.MapToSlice(espNamespaces, func(namespace string, _ struct{}) gatewayapiv1beta1.ReferenceGrantFrom {
+		return gatewayapiv1beta1.ReferenceGrantFrom{
+			Group:     egv1alpha1.GroupName,
+			Kind:      egv1alpha1.KindSecurityPolicy,
+			Namespace: gatewayapiv1.Namespace(namespace),
+		}
+	})
+	rg.Spec.From = refGrantFrom
+
+	return rg, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *EnvoySecurityPolicyReferenceGrantReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	ok, err := kuadrantenvoygateway.IsEnvoyGatewaySecurityPolicyInstalled(mgr.GetRESTMapper())
+	if err != nil {
+		return err
+	}
+	if !ok {
+		r.Logger().Info("Envoy SecurityPolicy ReferenceGrant controller disabled. EnvoyGateway API was not found")
+		return nil
+	}
+
+	ok, err = kuadrantgatewayapi.IsGatewayAPIInstalled(mgr.GetRESTMapper())
+	if err != nil {
+		return err
+	}
+	if !ok {
+		r.Logger().Info("Envoy SecurityPolicy ReferenceGrant controller disabled. GatewayAPI was not found")
+		return nil
+	}
+
+	securityPolicyToKuadrantEventMapper := mappers.NewSecurityPolicyToKuadrantEventMapper(
+		mappers.WithLogger(r.Logger().WithName("securityPolicyToKuadrantEventMapper")),
+		mappers.WithClient(r.Client()),
+	)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&kuadrantv1beta1.Kuadrant{}).
+		Owns(&gatewayapiv1beta1.ReferenceGrant{}).
```

Comment:
I cannot see the ownerref being set to the referencegrant

Reply:
Fixed this